### PR TITLE
feat: Add GitLabDocumentLoader for importing files from GitLab repositories

### LINKED ISSUE #4297: [FEATURE] Add GitLabDocumentLoader for GitLab Repositories
**Is your feature request related to a problem? Please describe.**
Yes. Currently, **LangChain4j** lacks a built-in mechanism to ingest source code and documentation directly from **GitLab** repositories.

GitLab is a standard tool for storing code and documentation in many enterprises. Developers building RAG (Retrieval-Augmented Generation) applications (e.g., "Chat with your Codebase") often need to index files stored in GitLab.

While LangChain4j already supports GitHub, there is no equivalent loader for GitLab, forcing developers to clone repositories locally or write custom API clients to fetch files.

**Describe the solution you'd like**
I would like to propose a new module `langchain4j-document-loader-gitlab` that implements a `GitLabDocumentLoader`.

The loader should leverage the **GitLab REST API v4** and support the following features:

1. **Dual Identity Support:** Support fetching from projects using both numeric Project IDs (e.g., `12345`) and URL-encoded Project Paths (e.g., `group/project`).
2. **Smart Branch Detection:** If a specific branch is not provided, it should automatically attempt to fetch from `main`, falling back to `master` if `main` does not exist.
3. **Authentication:** Support Personal Access Tokens via the `PRIVATE-TOKEN` header.
4. **Recursion & Pagination:** Support recursive file listing (`recursive=true`) and handle pagination automatically using the `X-Next-Page` header to ensure complete file tree traversal.
5. **Robust URL Handling:** Correctly handle URL encoding for file paths (e.g., encoding `/` as `%2F` and spaces as `%20` instead of `+`) to comply with GitLab API strictness.
6. **Metadata:** Automatically extract metadata such as `file_name`, `gitlab_project_id`, and the clickable web `url` for the file.

**Describe alternatives you've considered**

1. **Git Clone:** Cloning the repository to the local file system and using `FileSystemDocumentLoader`. This requires disk space and managing git credentials/SSH keys in the application environment.
2. **GitHub Loader:** Trying to adapt the existing GitHub loader, but the API endpoints and authentication mechanisms differ significantly.
3. **Custom Implementation:** Writing ad-hoc `HttpClient` logic for every project.

**Additional context**
I have already implemented a working prototype that uses `java.net.http.HttpClient` (standard JDK 11+) and `Jackson` for JSON parsing.

The implementation handles:

* Both GitLab SaaS (`gitlab.com`) and Self-Managed instances via configurable `baseUrl`.
* Filtering by `ref` (branch/tag) and sub-directory `path`.

I am ready to submit a Pull Request for this feature.

### PR DESCRIPTION
**Pull request by qicesun**
## Issue
Closes #4297

## Change
This PR introduces `langchain4j-document-loader-gitlab`, enabling seamless ingestion of documents and code from **GitLab** (SaaS & Self-Managed).

**Motivation:**
Following the recent addition of the `ConfluenceDocumentLoader`, this module completes the "Enterprise Knowledge" puzzle by supporting GitLab, ensuring developers can ingest both documentation and source code from their repositories.

**Key Features:**
* **Dual Identity Support:** Handles both **Project ID** (numeric e.g., `12345`) and **Project Path** (URL-encoded e.g., `group/project`) seamlessly for both API calls and Web URL generation.
* **Smart Branch Detection:** Automatically attempts to fetch from `main`, falling back to `master` if the default branch is not explicitly provided, simplifying configuration for older repositories.
* **Robust Encoding:** Specifically handles GitLab API's strict requirements for URL encoding (e.g., handling `%2F` in paths and ensuring `%20` for spaces instead of `+`).
* **Pagination:** Implements `do-while` loops respecting the `X-Next-Page` header to ensure complete file tree traversal.
* **Recursive Loading:** Leverages GitLab API's `recursive=true` parameter for efficient tree retrieval.

**Implementation Details:**
* Uses standard `java.net.http.HttpClient` to minimize external dependencies.
* Added extensive unit tests with `Mockito` covering pagination loops, branch fallback logic, and complex URL construction edge cases.

## General checklist
- [x] There are no breaking changes (API, behaviour)
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module
- [x] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by qicesun** [comment]
**Regarding the repository choice:**

I noticed the guideline suggests adding new integrations to the langchain4j-community repo.

However, I submitted this PR to the main repository because I saw that langchain4j-document-loader-github (and S3/Azure) are already located here. I assumed that core document loaders like GitLab/Confluence should ideally reside alongside them for consistency.

If you prefer this to be moved to the community repo, please let me know, and I will be happy to migrate it.

File: document-loaders/langchain4j-document-loader-gitlab/src/main/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoader.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+
+    public static final String METADATA_GITLAB_PROJECT_ID = "gitlab_project_id";
+
+    private static final Logger logger = LoggerFactory.getLogger(GitLabDocumentLoader.class);
+
+    private static final String DEFAULT_BASE_URL = "https://gitlab.com";
+    private static final int DEFAULT_PER_PAGE = 100;
+    private static final Duration DEFAULT_REQUEST_TIMEOUT = Duration.ofSeconds(30);
+
+    private final HttpClient httpClient;
```
Comment (dliubarskyi) [inline document-loaders/langchain4j-document-loader-gitlab/src/main/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoader.java:54]: Please use `dev.langchain4j.http.client.HttpClient` instead

File: document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+
+class GitLabDocumentLoaderTest {
+
+    @Test
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    void shouldLoadTwoDocumentsFromRepositoryTree() throws Exception {
+        HttpClient httpClient = mock(HttpClient.class);
```
Comment (dliubarskyi) [inline document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java:26]: You can use `dev.langchain4j.http.client.MockHttpClient`

File: document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+
+class GitLabDocumentLoaderTest {
```
Comment (dliubarskyi) [inline document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java:22]: Is there a way to test it against a real instance? Perhaps testcontainers?

**Review by dliubarskyi — COMMENTED** [review]
@qicesun thank you!

File: document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java (hunk 3 above)
Comment (qicesun) [reply document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java:22]: **Done**
Added `GitLabDocumentLoaderIT` uses a real GitLab instance (controlled by an environment variable), not Testcontainers, but it follows the same approach as other loaders

File: document-loaders/langchain4j-document-loader-gitlab/src/main/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoader.java (hunk 1 above)
Comment (qicesun) [reply document-loaders/langchain4j-document-loader-gitlab/src/main/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoader.java:54]: **Done**
Migrated All HTTP Client to dev.langchain4j.http.client

File: document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java (hunk 2 above)
Comment (qicesun) [reply document-loaders/langchain4j-document-loader-gitlab/src/test/java/dev/langchain4j/data/document/loader/gitlab/GitLabDocumentLoaderTest.java:26]: **Done**
Migrated All HTTP Client to dev.langchain4j.http.client

**Comment by dliubarskyi** [comment]
Could you please re-open this PR in the https://github.com/langchain4j/langchain4j-community repository?
